packages/ethereum-contracts/contracts/EIP712_Canvas.sol

```diff
@@ -0,0 +1,274 @@
+        // https://github.com/multiformats/multicodec/blob/696e701b6cb61f54b67a33b002201450d021f312/table.csv#L41
+        (digestSize,digest) = CID.createDigest(0x55, multihash);
+
+        return CID.encodeCID(1, 712, digest);
+    }
+
+    /**
+     * Hash a Session.
+     */
+    function hashSession(
+        Session memory session
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encode(
+            keccak256(bytes(sessionType)),
+            session.address_,
+            keccak256(bytes(session.blockhash)),
+            session.duration,
+            keccak256(bytes(session.publicKey)),
+            session.timestamp
+        ));
+    }
+
+    /**
+     * Hash an Action. It is expected that the action's `args` have already been hashed.
+     */
+    function hashAction(
+        Action memory action
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encode(
+            keccak256(bytes(actionType)),
+            action.address_,
+            keccak256(action.args),
+            keccak256(bytes(action.blockhash)),
+            keccak256(bytes(action.name)),
+            action.timestamp
+        ));
+    }
+
+    /**
+     * Hash a Message<Session>.
+     */
+    function hashSessionMessage(
+        SessionMessage memory sessionMessage
+    ) public pure returns (bytes32) {
+        return
+        keccak256(
+            abi.encode(
+                keccak256(bytes(sessionMessageType)),
+                sessionMessage.clock,
+                _hashStringArray(sessionMessage.parents),
+                hashSession(sessionMessage.payload),
+                keccak256(bytes(sessionMessage.topic))
+            )
+        );
+    }
+
+    /**
+     * Hash a Message<Action>.
+     */
+    function hashActionMessage(
+        ActionMessage memory actionMessage
+    ) public pure returns (bytes32) {
+        return
+        keccak256(
+            abi.encode(
+                keccak256(bytes(actionMessageType)),
+                actionMessage.clock,
+                _hashStringArray(actionMessage.parents),
+                hashAction(actionMessage.payload),
+                keccak256(bytes(actionMessage.topic))
+            )
+        );
+    }
+
+    /**
+     * Recover the address that signed a Session.
+     */
+    function recoverAddressFromSession(
+        Session memory session,
+        string memory name
+    ) public pure returns (address) {
+        bytes32 digest = _hashTypedDataV4(hashSession(session), name);
+
+        return ECDSA.recover(digest, session.authorizationData.signature);
+    }
+
+    /**
+     * Verify a Session.
+     */
+    function verifySession(
+        Session memory session,
+        address expectedAddress,
+        string memory name
+    ) public pure returns (bool) {
+        bytes32 digest = _hashTypedDataV4(hashSession(session), name);
+
+        return ECDSA.recover(digest, session.authorizationData.signature) == expectedAddress;
+    }
```

Comment:
verifySessionWithPubkey

Reply:
tempted to not implement this until we actually have some code that needs it